File: app/src/main/java/phone/vishnu/todoapp/repository/ShelveRepository.java
package phone.vishnu.todoapp.repository;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import java.util.List;

import phone.vishnu.todoapp.dao.ShelveDao;
import phone.vishnu.todoapp.database.ShelveDatabase;
import phone.vishnu.todoapp.model.Shelve;

public class ShelveRepository {

    private ShelveDao shelveDao;
    private LiveData<List<Shelve>> shelvesList;

    public ShelveRepository(Application application) {
        ShelveDatabase database = ShelveDatabase.getInstance(application);
        shelveDao = database.shelveDao();
        shelvesList = shelveDao.getAllShelves();
    }

    public void insertShelve(Shelve shelve) {
        new InsertShelveAsyncTask(shelveDao).execute(shelve);
    }

    public void updateShelve(Shelve shelve) {
        new UpdateShelveAsyncTask(shelveDao).execute(shelve);
    }

    public void deleteShelve(Shelve shelve) {
        new DeleteShelveAsyncTask(shelveDao).execute(shelve);
    }

    public LiveData<List<Shelve>> getAllShelves() {
        return shelvesList;
    }

    private static class InsertShelveAsyncTask extends AsyncTask<Shelve, Void, Void> {
        private ShelveDao shelveDao;

        public InsertShelveAsyncTask(ShelveDao shelveDao) {
            this.shelveDao = shelveDao;
        }

        @Override
        protected Void doInBackground(Shelve... shelves) {
            shelveDao.insert(shelves[0]);
            return null;
        }
    }

    private static class UpdateShelveAsyncTask extends AsyncTask<Shelve, Void, Void> {
        private ShelveDao shelveDao;

        public UpdateShelveAsyncTask(ShelveDao shelveDao) {
            this.shelveDao = shelveDao;
        }

        @Override
        protected Void doInBackground(Shelve... shelves) {
            shelveDao.update(shelves[0]);
            return null;
        }
    }

    private static class DeleteShelveAsyncTask extends AsyncTask<Shelve, Void, Void> {
        private ShelveDao shelveDao;

        public DeleteShelveAsyncTask(ShelveDao shelveDao) {
            this.shelveDao = shelveDao;
        }

        @Override
        protected Void doInBackground(Shelve... shelves) {
            shelveDao.delete(shelves[0]);
            return null;
        }
    }
}

File: app/src/main/java/phone/vishnu/todoapp/repository/ShelveRepository.kt
package phone.vishnu.todoapp.repository

import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import phone.vishnu.todoapp.dao.ShelveDao
import phone.vishnu.todoapp.model.Shelve

class ShelveRepository(private val shelveDao: ShelveDao) {

    private val shelvesList: LiveData<List<Shelve>> = shelveDao.getAllShelves()

    @WorkerThread
    suspend fun insert(shelve: Shelve) {
        shelveDao.insert(shelve)
    }

    @WorkerThread
    suspend fun delete(shelve: Shelve) {
        shelveDao.delete(shelve)
    }

    @WorkerThread
    suspend fun update(shelve: Shelve) {
        shelveDao.update(shelve)
    }

    fun getAllShelves(): LiveData<List<Shelve>> {
        return shelvesList
    }
}